### ISSUE TITLE
Screenshots sometimes don't include background images

### ISSUE THREAD

**Issue by trotzig** [mozilla/geckodriver#367]
I'm using [selenium-webdriver](https://github.com/SeleniumHQ/selenium/tree/master/javascript/node/selenium-webdriver) to take screenshots in [a visual diffing tool called happo](https://github.com/Galooshi/happo). Up until now, we've been using the firefox driver previously shipped with selenium. Since that driver is no longer maintained, we're looking into switching to geckodriver. A nice side-effect from that switch is that we'll be able to screenshot individual elements instead of full screens (which is why I filed https://github.com/mozilla/geckodriver/issues/366 asking for a new release). 

After building and using the latest trunk version of geckodriver, I've noticed a difference from the old firefoxdriver: screenshots now may or may not include background images. Here's a resulting [happo](https://github.com/Galooshi/happo) diff showing the difference between two different screenshots of an element twice:

<img width="697" alt="screen shot 2016-11-27 at 19 57 03" src="https://cloud.githubusercontent.com/assets/578277/20650965/bad25bda-b4db-11e6-8647-5bf03cd975d6.png">

(also available at https://diffux-trotzig.s3.amazonaws.com/a0aaaf2aa5719cc56796680047f00ea8/index.html)

I can't recall seeing that happen in the old firefox driver. 

If you don't agree with me in considering this a bug/issue, it would be helpful to have a discussion around how to implement something that waits for background images before taking a screenshot. Regular images are usually fine, and [we already have something in place to wait for them](https://github.com/Galooshi/happo/blob/5c5550a240d8db4b83dfe479b4290946f7c17786/js/src/waitForImagesToRender.js). But background images are trickier. Maybe the only way to do it is to recursively check `getComputedStyle` for all elements and look for `url`s?

**Comment by AutomatedTester** [comment]
if this is a case of waiting for the files to be available then maybe. If it's added via JS then its out of scope, if it's part of the original markup then let's look into it further.

Can you give a [concise reproducible test case](http://sscce.org/).

**Comment by trotzig** [comment]
Yeah, this is added via javascript. I could produce a minimal repro case, but if it's out of scope, I won't bother. I wonder what's different though. I can't find anything in the [firefoxdriver source](https://github.com/SeleniumHQ/selenium/blob/ceaf3da79542024becdda5953059dfbb96fb3a90/javascript/firefox-driver/js/screenshooter.js#L23) that hints at what's different. It might just be that the old driver is slower so that images are available before we take the screenshot.

**Comment by andreastt** [comment]
Marionette is meant to wait for the document [readyState](https://developer.mozilla.org/en/docs/Web/API/Document/readyState) to reach "complete". If the background image is loaded via JS following that, you have to write a wait condition to ensure it is loaded fully before using the take screenshot primitive.

**Comment by andreastt** [comment]
Something like this might help:

```html
<script>
let img = document.querySelector("#my-background-image");
img.addEventListener("load", () => window.fullyLoaded = true, {once: true});
</script>
```

WebDriver pseudo code:

```py
session.navigate("https://my.domain")
Wait(session).until(lambda s: s.execute_script("return window.fullyLoaded"))
capture = session.take_screenshot()
```

**Comment by trotzig** [comment]
Thanks for the suggestion @andreastt, I worried I would have to do something like that. The problem is we don't know all background images before (happo is a general tool) so we would have to introduce a new responsibility for consumers of our tool to somehow list all background images in config. With regular images, we can just query the DOM and dynamically preload all `src`es. But for background images it gets more complicated. 

I appreciate the quick responses!

**Comment by trotzig** [comment]
Wait, I think I read your response too quickly... Does the `load` event propagate to its parents? In that case we can just attach a listener to root nodes after we inject them.

**Comment by andreastt** [comment]
Unfortunately the `load`/`onload` event does not bubble, but you would have to enumerate all your `<img>` elements first to know what to wait for anyway.

**Comment by lencioni** [comment]
Is there an API that we could use to observe any pending HTTP requests?

**Comment by trotzig** [comment]
Closing this to reduce noise, although I'm still interested in hearing an answer to @lencioni's question about HTTP requests above. 

The workaround we're applying involves grabbing all elements from the document and using `getComputedStyle` to look for possible background images, then preload those. Something like this:

```js
function waitForImageToLoad(url) {
  return new Promise((resolve) => {
    const img = new Image();
    img.onload = resolve;
    img.src = url;
  });
}
const promises = [];
document.body.querySelectorAll('*').forEach((element) => {
  const computedStyle = window.getComputedStyle(element);
  const bgImage = computedStyle.getPropertyValue('background-image');
  if (bgImage && bgImage.startsWith('url(')) {
    const url = bgImage.replace(/^url\(['"]/, '').replace(/['"]?\)$/, '');
    promises.push(waitForImageToLoad(url));
  }
});
```

**Comment by andreastt** [comment]
There’s no WebDriver API to observe HTTP requests. Determining if any requests are ‘pending’ isn’t possible since they may occur at any time.

@trotzig That’s an elegant solution. From the WebDriver client you could potentially do an explicit wait on all the promises to resolve before taking the screenshot.

**Comment by trotzig** [comment]
Yep, that's what we'll do. Thanks for the feedback.

**Comment by andreastt** [comment]
@trotzig Just a comment on one of the lines you pasted:

```js
    img.onload = resolve;
```

This will attach the callback forever, which might not be an issue since most images don’t load more than once. But you might want to use `img.addEventListener("load", resolve, {once: true})` so that the event listener is removed implicitly the first time it hits.

**Comment by trotzig** [comment]
Great, thanks for the suggestion!
